### ISSUE TITLE
Authorization fails in procedure calls

### ISSUE THREAD

**Issue by colinmackinlay** [livewire/livewire#588]
When calling $this->authorize as per the docs I get an unauthorized 403 error.

The policy called (with added logging) is:
```
    public function update(Staff $user, RiskAssessment $assessment)
    {
        info(1);
        if ($user->can('riskassess_approve')) return true;
        info(2);
        if ($assessment->owner_id == $user->id) return true;
        info(3);
        return false;
    }
```
In running livewire tests I get the 1 logged and 2 which at least indicates that it is called and returns the correct answer. In the browser though, nothing is logged and I get the 403.

The livewire component seems to have access to auth() in the mounted() procedure but not in other function calls when I call from the browser rather than a test. To confirm this I've added some logging of auth()->user()->name:

```
class RiskAssessmentView extends Component {

    use AuthorizesRequests;

    public $assessment;

    public function mount(RiskAssessment $assessment)
    {
        $this->assessment = $assessment;
        info(auth()->user()->name);
    }

    public function deleteAction($action_id)
    {
        info(auth()->user()->name);

        $this->authorize('update', $this->assessment);

        if (RiskAction::find($action_id)->risk_assessment_id != $this->assessment->id) return;

        RiskAction::destroy($action_id);
    }
```

The result is:
```
[2020-02-04 20:35:59] local.INFO: Test Admin
[2020-02-04 20:36:13] local.ERROR: Trying to get property 'name' of non-object 
```
The first line is logged from mount(), the second comes when I click a component in the blade to call the deleteAction() using:
```wire:click.prevent="deleteAction('{{$risk_action->id}}')"```
In other words I can't get authorize to work in a browser

This is driving me potty. Any ideas very welcome. Seems possible it's a livewire issue but the difference between test and browser troubles me!

I**Desktop (please complete the following information):**
 - Browser Chrome, Apache 2.4.35, php 7.4.2

**Comment by colinmackinlay** [comment]
Further investigation makes it seem even more like a livewire issue. In all my livewire components I have checked for availability of auth()->user() using:

```        ddd(auth()->user());```

If I do that in mount() or in render() then I can see the user details but in any called method this returns null. 

That explains why authorize() will not work in any of the called methods.

**Comment by calebporzio** [comment]
Livewire uses the `web` middleware by default for subsequent requests. Which by default includes the `StartSession::class` middleware that is in charge of providing things like `auth()->user()`.

![image](https://user-images.githubusercontent.com/3670578/73956236-a82bdb80-48d2-11ea-937c-2ea239d06afd.png)

I'm guessing you have some kind of custom setup that is a little different.

If that is the case, you can configure the middleware group that Livewire uses in the config file:

`php artisan vendor:publish --tag=livewire:config `

Then this is the key:
![image](https://user-images.githubusercontent.com/3670578/73956379-ddd0c480-48d2-11ea-8d36-fa52aaa6830c.png)

Hope this helps!

**Comment by colinmackinlay** [comment]
Thanks @calebporzio for taking a look. I've not got any particularly custom setup and "web" is standard on all routes. The top of the stack when I'm debugging is below and shows StartSession being called
```
Gate.php:331, Illuminate\Auth\Access\Gate->inspect()
Gate.php:318, Illuminate\Auth\Access\Gate->authorize()
AuthorizesRequests.php:23, App\Http\Livewire\RiskAssessmentView->authorize()
RiskAssessmentView.php:23, App\Http\Livewire\RiskAssessmentView->deleteAction()
HandlesActions.php:88, App\Http\Livewire\RiskAssessmentView->callMethod()
ConnectionHandler.php:51, Livewire\Connection\HttpConnectionHandler->processMessage()
ConnectionHandler.php:21, Livewire\Connection\HttpConnectionHandler->handle()
HttpConnectionHandler.php:11, Livewire\Connection\HttpConnectionHandler->__invoke()
ControllerDispatcher.php:48, Illuminate\Routing\ControllerDispatcher->dispatch()
Route.php:219, Illuminate\Routing\Route->runController()
Route.php:176, Illuminate\Routing\Route->run()
Router.php:680, Illuminate\Routing\Router->Illuminate\Routing\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Routing\Router.php:678-682}()
Pipeline.php:130, Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php:128-136}()
CreateFreshApiToken.php:40, Laravel\Spark\Http\Middleware\CreateFreshApiToken->handle()
Pipeline.php:171, Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php:147-180}()
SubstituteBindings.php:41, Illuminate\Routing\Middleware\SubstituteBindings->handle()
Pipeline.php:171, Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php:147-180}()
VerifyCsrfToken.php:76, App\Http\Middleware\VerifyCsrfToken->handle()
Pipeline.php:171, Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php:147-180}()
ShareErrorsFromSession.php:49, Illuminate\View\Middleware\ShareErrorsFromSession->handle()
Pipeline.php:171, Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php:147-180}()
StartSession.php:56, Illuminate\Session\Middleware\StartSession->handle()
Pipeline.php:171, Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php:147-180}()
AddQueuedCookiesToResponse.php:37, Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle()
```
Following the debugging through a step at a time it never gets to my authorize method when run in a browser.
When run as a test it works fine:
```
RiskAssessmentView.php:25, App\Http\Livewire\RiskAssessmentView->deleteAction()
HandlesActions.php:88, App\Http\Livewire\RiskAssessmentView->callMethod()
ConnectionHandler.php:51, Livewire\Connection\HttpConnectionHandler->processMessage()
ConnectionHandler.php:21, Livewire\Connection\HttpConnectionHandler->handle()
HttpConnectionHandler.php:11, Livewire\Connection\HttpConnectionHandler->__invoke()
ControllerDispatcher.php:48, Illuminate\Routing\ControllerDispatcher->dispatch()
Route.php:219, Illuminate\Routing\Route->runController()
Route.php:176, Illuminate\Routing\Route->run()
Router.php:680, Illuminate\Routing\Router->Illuminate\Routing\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Routing\Router.php:678-682}()
Pipeline.php:130, Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php:128-136}()
Pipeline.php:105, Illuminate\Routing\Pipeline->then()
Router.php:682, Illuminate\Routing\Router->runRouteWithinStack()
Router.php:657, Illuminate\Routing\Router->runRoute()
Router.php:623, Illuminate\Routing\Router->dispatchToRoute()
Router.php:612, Illuminate\Routing\Router->dispatch()
Kernel.php:176, App\Http\Kernel->Illuminate\Foundation\Http\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php:173-177}()
Pipeline.php:130, Illuminate\Routing\Pipeline->Illuminate\Pipeline\{closure:C:\Users\Colin\Websites\CSS\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php:128-136}()
Pipeline.php:105, Illuminate\Routing\Pipeline->then()
Kernel.php:151, App\Http\Kernel->sendRequestThroughRouter()
Kernel.php:116, App\Http\Kernel->handle()
MakesHttpRequests.php:375, Livewire\Testing\MakesHttpRequestsWrapper->call()
TestableLivewire.php:111, Livewire\Testing\TestableLivewire->Livewire\Testing\{closure:C:\Users\Colin\Websites\CSS\vendor\livewire\livewire\src\Testing\TestableLivewire.php:103-113}()
MakesHttpRequestsWrapper.php:28, Livewire\Testing\MakesHttpRequestsWrapper->temporarilyDisableExceptionHandlingAndMiddleware()
TestableLivewire.php:113, Livewire\Testing\TestableLivewire->pretendWereSendingAComponentUpdateRequest()
MakesCallsToComponent.php:66, Livewire\Testing\TestableLivewire->sendMessage()
MakesCallsToComponent.php:30, Livewire\Testing\TestableLivewire->runAction()
MakesCallsToComponent.php:24, Livewire\Testing\TestableLivewire->call()
RiskActionTest.php:27, Tests\Feature\RiskActionTest->it_displays_a_risk_action()
TestCase.php:1154, Tests\Feature\RiskActionTest->runTest()
TestCase.php:842, Tests\Feature\RiskActionTest->runBare()
TestResult.php:693, PHPUnit\Framework\TestResult->run()
TestCase.php:796, Tests\Feature\RiskActionTest->run()
TestSuite.php:746, PHPUnit\Framework\TestSuite->run()
TestRunner.php:652, PHPUnit\TextUI\TestRunner->doRun()
Command.php:206, PHPUnit\TextUI\Command->run()
Command.php:162, PHPUnit\TextUI\Command::main()
phpunit:61, {main}()
```
In fact the stack here is after we've been through the authorize method three times in the test:
```
        $assessment = RiskAssessmentFactory::ownedBy($this->user)->withRiskActions(1)->create();
        $action_id = $assessment->risk_actions->first()->id;
        Livewire::actingAs($this->user)
            ->test(RiskAssessmentView::class, $assessment)
            ->assertSee('Risk Assessment')
            ->assertDontSee('New action')
            ->call('createAction')
            ->assertSee('New action')
            ->call('deleteAction', $action_id)
            ->assertDontSee('New action');
```
Any other magic thoughts? :)

**Comment by calebporzio** [comment]
Ah, so are you only seeing the issue from a test? Like the subsequent requests work fine from the browser?

**Comment by colinmackinlay** [comment]
No, other way around. In the test, authorize gets called fine and behaves correctly but in a browser it never hits it. I can follow the debug through to it being called but laravel stops it getting as far as the policy method because there is no authorised user. Sorry I wasn't clear.

I'm currently building a fresh laravel instance with a simple example to check whether it happens in that.

**Comment by colinmackinlay** [comment]
Update - a fresh Laravel install with just a Livewire component with a model policy works fine so it clearly must be something with my main app. I'll close the issue now and report back when I've tracked the issue down in case its something that might happen to others.

**Comment by colinmackinlay** [comment]
Not tracked down completely but pretty sure this is actually to do with me using tenancy/multi-tenant.
I have now successfully setup a livewire component using authorisation which works fine if the route is provided through the top level host but when the routes go through tenant sub-domains auth()->user() only returns correctly on instantiation/mount of livewire but on procedure calls it returns null. I may try to dig further but I'm hoping that @luceos will want to ensure livewire works  with tenancy :)

**Comment by calebporzio** [comment]
Gotcha, must be something different that `tenancy/multi-tenant` does with the `web`middleware group (the default group applied to subsequent Livewire requests).

If you find something out. Re-open and we can talk solutions. Thanks for following up!

**Comment by colinmackinlay** [comment]
Hi @calebporzio I've reproduce the issue in a clean install of laravel and livewire (which works fine) but with the skeleton of multitenant added in:
https://github.com/colinmackinlay/livewire

I've written a commentary as the readme which follows through the commits. There's probably some unnecessary code as I've had to hack it around to leave Laravel Spark out which was generating the tenants. Although it's messy it isn't doing anything weird with the web middleware group

The only thing that I can think could be an issue is that for some reason the fact that I'm using a different guard for my authenticated tenant users could be interfering with livewire.

Logging of auth()->user() at different places in the Livewire Component shows that it's not present when a method is called by a wire:click event

**Comment by colinmackinlay** [comment]
I've tested whether it's just using another guard in a separate branch from before multitenant
https://github.com/colinmackinlay/livewire/tree/other_guard
That works fine though. 
(I did need to update my render method to call fresh() on the local copy stored during mount - it makes sense that changes in the database won't flow through to that but strangely, they did in the original commit)

So now we get my original behaviour issue which is that: I can pass the tests with authorisation working fine but not in the browser. Annotated log file below:
```
/country/4 - Tenant is recognised as auth()->user() on mount and render
 [2020-02-22 10:48:55] local.INFO: Auth user when mounted: Tenant
 [2020-02-22 10:48:55] local.INFO: Auth user in render method: Tenant
/click the delete link on heathrow - no auth()->user(0 found
 [2020-02-22 10:49:09] local.INFO: Request delete airport with id 1
 [2020-02-22 10:49:09] local.INFO: Auth user in deleteAirport method: null
                                                                      ^^^^
/now run tests again
 [2020-02-22 10:51:20] testing.INFO: Auth user when mounted: null  
 [2020-02-22 10:51:20] testing.INFO: Auth user in render method: null  
 [2020-02-22 10:51:20] testing.INFO: Auth user when mounted: null  
 [2020-02-22 10:51:20] testing.INFO: Auth user in render method: null  
 [2020-02-22 10:51:20] testing.INFO: Auth user when mounted: null  
 [2020-02-22 10:51:20] testing.INFO: Auth user in render method: null  
 [2020-02-22 10:51:20] testing.INFO: Request delete airport with id 1  
 [2020-02-22 10:51:20] testing.INFO: Auth user in deleteAirport method: null  
 [2020-02-22 10:51:20] testing.INFO: Auth user when mounted: Mohammed Considine  
 [2020-02-22 10:51:20] testing.INFO: Auth user in render method: Mohammed Considine  
 [2020-02-22 10:51:20] testing.INFO: Request delete airport with id 1  
/ auth()->user() is found in the delete method during tests
 [2020-02-22 10:51:20] testing.INFO: Auth user in deleteAirport method: Mohammed Considine  
                                                                        ^^^^^^^^^^^^^^^^^^
 [2020-02-22 10:51:20] testing.INFO: Policy authorisation requested  
 [2020-02-22 10:51:20] testing.INFO: Delete airport with id 1  
 [2020-02-22 10:51:20] testing.INFO: Auth user in render method: Mohammed Considine
```

**Comment by colinmackinlay** [comment]
I've revisited this with a much simpler setup:
Laravel 7 + Livewire 1.x + Sushi - all works fine

Adding hyn/multi-tenant (5.6.3) which draws in:
  - phpseclib/phpseclib (2.0.27)
  - doctrine/event-manager (1.1.0)
  - doctrine/dbal (v2.10.1)
  - doctrine/cache (1.10.0)
and just configuring enough to get it running (a system MySql database) means the problem appears. I don;t even need to move the livewire component models into a tenant.

I assume it must be to do with one of the packages above and something hyn/multi-tenant is doing at a very basic level. 

Given that hyn/multi-tenant is not being developed as much as its replacement then it seems to time to research and jump ship to another SaaS tenancy system without the problems :(

**Comment by colinmackinlay** [comment]
Actually - the issue is the other change when going multi-tenant in that the authenticated user I have uses a different guard. All this is fixed by adding the new guard to the livewire config:
```
    'middleware_group' => ['web', 'auth:otherguard'],

```

**Comment by AhmedSewidan** [comment]
<?php

namespace Modules\Core\Traits\Dashboard;

use Livewire\Attributes\On;
use Spatie\Permission\Models\Permission;

trait HasPermissions
{
    public array $guards = ['web', 'admin', 'doctor'];
    public array $authUsers = [];

    /**
     * Cache authenticated users from all guards.
     * Should be used in mount
     */
    public function storeUserProperties(): void
    {
        foreach ($this->guards as $guard) {
            if ($user = auth($guard)->user()) {
                $this->authUsers[$guard] = $user;
            }
        }
    }

    /**
     * Restore cached authenticated users into their respective guards.
     * Should be used in boot
     */
    public function restoreAuthUsers(): void
    {
        foreach ($this->authUsers as $guard => $user) {
            auth($guard)->setUser($user);
        }
    }

    /**
     * Refresh permissions for all authenticated users.
     */
    #[On('refresh-permissions')]
    public function refreshPermissions(): void
    {
        foreach ($this->authUsers as $user) {
            if ($user->relationLoaded('roles')) {
                $user->loadMissing('roles.permissions');
            }
        }

        Permission::forgetCachedPermissions();
    }
}


in Your Dynamic component do this    =>  ✅

<?php

namespace Modules\Core\Livewire\Dashboard;

use Modules\Core\Traits\Dashboard\HasPermissions;

class BaseTableComponent extends Component
{
    use HasPermissions;

    public function mount()
    {
        $this->storeUserProperties();
    }

    public function boot()
    {
        $this->restoreAuthUsers();
    }

    //  Your code here ....
}

**Comment by AhmedSewidan** [comment]
```
Best solution ✅✅✅✅

namespace Modules\Core\Traits\Dashboard;

use Livewire\Attributes\On;
use Spatie\Permission\Models\Permission;

trait HasPermissions
{
    public array $guards = ['web', 'admin', 'doctor'];
    public array $authUsers = [];

    /**
     * Cache authenticated users from all guards.
     * Should be used in mount
     */
    public function storeUserProperties(): void
    {
        foreach ($this->guards as $guard) {
            if ($user = auth($guard)->user()) {
                $this->authUsers[$guard] = $user;
            }
        }
    }

    /**
     * Restore cached authenticated users into their respective guards.
     * Should be used in boot
     */
    public function restoreAuthUsers(): void
    {
        foreach ($this->authUsers as $guard => $user) {
            auth($guard)->setUser($user);
        }
    }

    /**
     * Refresh permissions for all authenticated users.
     */
    #[On('refresh-permissions')]
    public function refreshPermissions(): void
    {
        foreach ($this->authUsers as $user) {
            if ($user->relationLoaded('roles')) {
                $user->loadMissing('roles.permissions');
            }
        }

        Permission::forgetCachedPermissions();
    }
}

in Your Dynamic component do this    =>  ✅

class BaseTableComponent extends Component
{
    use HasPermissions;

    public function mount()
    {
        $this->storeUserProperties();
    }

    public function boot()
    {
        $this->restoreAuthUsers();
    }

    //  Your code here ....
}
```
